# Installation and test run

1. install miniconda
	```
	wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
	sudo bash Miniconda3-latest-Linux-x86_64.sh -b -p /usr/local/miniconda3
	echo "export PATH=/usr/local/miniconda3/bin:$PATH" >> ~/.bashrc
	source ~/.bashrc
	sudo chmod -R 775 /usr/local/miniconda3/
	```
2.  update conda
	```
	conda upgrade  conda -y
	```
3.  add conda channels
	```
	conda config --add channels https://conda.anaconda.org/dranew
	conda config --add channels 'bioconda'
	conda config --add channels 'r'
	conda config --add channels 'conda-forge'
	conda config --add channels https://conda.anaconda.org/aroth85
	conda config --add channels https://conda.anaconda.org/shahcompbio
	```
4.  Install single cell pipeline
	```
	conda install -c shahcompbio single_cell_pipeline
	```
7. download test data
	```
        wget https://singlecelldata.blob.core.windows.net/dlp-paper/data.tar.gz
        sudo mkdir /data
        sudo tar -xvf data.tar.gz -C /
	```
        *If you don't have permissions to write to /data directory, please feel free to extract it someplace else. see the note in step 10*
8. Download reference data:
	```
	wget https://singlecelldata.blob.core.windows.net/dlp-paper/refdata.tar.gz
	sudo mkdir /refdata
	sudo tar -xvf refdata.tar.gz -C /
	```
	*If you don't have permissions to write to /refdata directory, please feel free to extract it someplace else. see the note in step 12*
9. build pipeline input file.
	   The pipeline takes a yaml formatted input file with paths to fastq files, aligned bam files (will be generated by the pipeline) and metadata. The input yaml file structure is as follows:
	```
	SA123:
	  bam: data/SA123.bam
	  column: 1
	  condition: A
	  fastqs:
	    LANE_1:
	      fastq_1: data/SA123_R1.fastq.gz
	      fastq_2: data/SA123_R2.fastq.gz
	      sequencing_center: BCCAGSC
	      sequencing_instrument: HX
	  img_col: 2
	  index_i5: i5-1
	  index_i7: i7-2
	  pick_met: A1
	  primer_i5: AACCGGTT
	  primer_i7: ACGTACGT
	  row: 3
	  sample_type: null
	SA456:
	  bam: data/SA456.bam
	  column: 4
	  condition: A
	  fastqs:
	    LANE_1:
	      fastq_1: data/SA456_R1.fastq.gz
	      fastq_2: data/SA456_R2.fastq.gz
	      sequencing_center: BCCAGSC
	      sequencing_instrument: HX
	  img_col: 5
	  index_i5: i5-1
	  index_i7: i7-2
	  pick_met: B1
	  primer_i5: AACCGGTT
	  primer_i7: ACGTACGT
	  row: 6
	  sample_type: null
	``` 
        A sample input yaml file for the data downloaded in step 8 can be found in INSTALL/input.yaml.
        NOTE: The paths to files in INSTALL/input.yaml point to the /data directory. If you extracted the files to another location, please update the file accordingly.
10. generate the sample pipeline config file
	```
	single_cell generate_config --pipeline_config config.yaml
	```
	NOTE: If reference data is not in /refdata, then please update the following values to the correct path in the config.yaml file
	```gc_windows: /refdata/gc_windows.txt```
	```ref_genome: /refdata/GRCh37-lite.fa```
	```
    hmmcopy_params:
	  autoploidy:
        exclude_list: /refdata/repeats.satellite.regions
	```
	```
    hmmcopy_params:
      autoploidy:
        gc_wig_file: /refdata/GRCh37-lite.gc.ws_500000.wig
	```
	```
    hmmcopy_params:
      autoploidy:
        map_wig_file: /refdata/GRCh37-lite.map.ws_125_to_500000.wig
	```
11. run the alignment pipeline:
	```
	single_cell alignment --input_yaml single_cell_pipeline/INSTALL/input.yaml --out_dir dlp --bams_dir bams --submit local --library_id A96139A --loglevel DEBUG --tmpdir temp --pipelinedir pipeline --maxjobs 2 --config_file /path/to/config.yaml
	```
	Please refer to [doc](../../README.md) for detailed instructions for running all single cell sub commands.

	Replace /path/to/ with your path to config.yaml
    *Outputs:*
    * ```data/SA123.bam``` and ```data/SA456.bam```: aligned bam files
    * ```dlp/results/alignment/A123456_alignment_metrics.csv.gz```: alignment metrics and metadata
    * ```dlp/results/alignment/A123456_gc_metrics.csv.gz```: alignment metrics and metadata
    * ```dlp/results/alignment/plots/A123456_plot_metrics.pdf```: alignment QC plots.
	* ```dlp/results/hmmcopy_autoploidy/A123456_reads.csv.gz``` with data from hmmcopy
	* ```dlp/results/hmmcopy_autoploidy/A123456_metrics.csv.gz``` with data from hmmcopy
	* ```dlp/results/hmmcopy_autoploidy/A123456_params.csv.gz``` with data from hmmcopy
	* ```dlp/results/hmmcopy_autoploidy/A123456_segments.csv.gz``` with data from hmmcopy
    * ```dlp/results/hmmcopy_autoploidy/A123456_igv_segments.seg``` to load segments in IGV
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_heatmap_by_ec.pdf ```: copy number heatmap
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_heatmap_by_ec_filtered.pdf ```: heatmap with filters to remove bad cells
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_kernel_density.pdf ```: kernel density plot
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_metrics.pdf ``` hmmcopy metrics
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_bias.tar.gz ``` GC bias plots
    * ```dlp/results/hmmcopy_autoploidy/plots/A123456_segments.tar.gz ``` hmmcopy segments
