---
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document
params:
  output_name:
    value: x
---

title_var: "`r commandArgs(trailingOnly=T)[1]`"


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, fig.align = 'center')
library(knitr)
library(DT)
library(tidyr)
library(dplyr)
library(ggplot2)

```

```{r, include=FALSE}
#sample_id
#title_var: "`r commandArgs(trailingOnly=T)[1]`"

#title_var <- snakemake@params["sample_id"]
```
---
title: `r title_var`
---

# Cells per datatype
```{r datatype, echo = FALSE}
# All defaults
#abs datatype
read.csv(snakemake@params[["abs_datatype_summary"]]) %>%
  knitr::kable()
```

# SNV
```{r printsummary, echo = FALSE}
# All defaults
df <- read.csv(snakemake@params[["abs_summary"]])
```
Total number of cells: `r df$ncells`.

Total number of mutations: `r df$nmutations`.

## SNV cell counts + alt counts
![](`r snakemake@params[["abs_snvcellcounts"]]`) ![](`r snakemake@params[["abs_snvaltcounts"]]`)

## SNVs per cell
![](`r snakemake@params[["abs_muts_per_cell"]]`)

Number of mutations per chromosome:
```{r perchr, fig.height = 3, fig.width = 10, echo = FALSE}
# All defaults
snvs <- read.csv(snakemake@params[["abs_allsnvs"]])
chridx <- data.frame(chrom = c(paste0(1:22), "X", "Y"), idx = seq(1:24))
snvs %>%
  left_join(., chridx) %>%
  group_by(chrom, idx) %>%
  summarize(n = n()) %>%
  ggplot(aes(x = forcats::fct_reorder(chrom, idx), y = n)) +
  geom_bar(stat = "identity", fill = "plum4", width = 0.75) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  xlab("Chromosome") + ylab("Counts") + theme_classic()
```

## Rainfall plot
```{r adjdist, echo = FALSE}
# All defaults
include_graphics(snakemake@params[["abs_adj_distance"]])
```

## SNV density
```{r genomecount, echo = FALSE}
# All defaults
include_graphics(snakemake@params[["abs_genome_count"]])
```

## Mutation signatures
```{r, fig.height = 7, fig.width = 10, echo = FALSE}
library(ggplot2)
library(stringr)
library(dplyr)
library(tidyr)

df <- readr::read_csv(snakemake@params[["abs_trinuc"]])

sigs <- data.frame(trinuc_context = c(
    "ACAA", "ACAC", "ACAG", "ACAT", "CCAA", "CCAC", "CCAG", "CCAT",
    "GCAA", "GCAC", "GCAG", "GCAT", "TCAA", "TCAC", "TCAG", "TCAT",
    "ACGA", "ACGC", "ACGG", "ACGT", "CCGA", "CCGC", "CCGG", "CCGT",
    "GCGA", "GCGC", "GCGG", "GCGT", "TCGA", "TCGC", "TCGG", "TCGT",
    "ACTA", "ACTC", "ACTG", "ACTT", "CCTA", "CCTC", "CCTG", "CCTT",
    "GCTA", "GCTC", "GCTG", "GCTT", "TCTA", "TCTC", "TCTG", "TCTT",
    "ATAA", "ATAC", "ATAG", "ATAT", "CTAA", "CTAC", "CTAG", "CTAT",
    "GTAA", "GTAC", "GTAG", "GTAT", "TTAA", "TTAC", "TTAG", "TTAT",
    "ATCA", "ATCC", "ATCG", "ATCT", "CTCA", "CTCC", "CTCG", "CTCT",
    "GTCA", "GTCC", "GTCG", "GTCT", "TTCA", "TTCC", "TTCG", "TTCT",
    "ATGA", "ATGC", "ATGG", "ATGT", "CTGA", "CTGC", "CTGG", "CTGT",
    "GTGA", "GTGC", "GTGG", "GTGT", "TTGA", "TTGC", "TTGG", "TTGT"
)) %>%
  mutate(idx = 1:n()) %>%
  mutate(channel = paste0(str_sub(trinuc_context, 2,2), ">", str_sub(trinuc_context, 3,3))) %>%
  mutate(trinuc_contextplot = paste0(str_sub(trinuc_context,1 ,2),
                                    str_sub(trinuc_context, 4, 4)))

df <- df %>%
  mutate(trinuc_context = paste0(str_sub(tri_nucleotide_context,1 ,2),
                                 alt, str_sub(tri_nucleotide_context,3, 3))) %>%
  full_join(., sigs) %>%
  complete(num_cells_class, nesting(trinuc_context, idx, channel, tri_nucleotide_context, ref, alt)) %>%
  filter(!is.na(num_cells_class)) %>%
  mutate(n = ifelse(is.na(n), 0, n))
df$num_cells_class <- factor(df$num_cells_class, levels = c("1", "2-5", "6-20", ">20"))

df %>%
  ggplot(aes(x = idx, y = n, fill = channel)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("#1EBFF0", "#050708", "#E62725", "#CBCACB", "#A1CF64", "#EDC8C5")) +
  theme_classic() +
  scale_x_continuous(labels = sigs$trinuc_contextplot, breaks = sigs$idx) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5)) +
  xlab("") +
  ylab("Number of mutations") +
  guides(fill=guide_legend(title="")) +
  facet_wrap(~num_cells_class, ncol = 1, scales = "free_y")
```

## High impact + non-synonymous mutations
```{r, echo = FALSE}
df <- read.csv(snakemake@params[["abs_highimpact"]])
subset(df, select = -c(X, max_museq_score, max_strelka_score)) %>%
    mutate(gene_name = paste0('<a href="',"https://www.genecards.org/cgi-bin/carddisp.pl?gene=",gene_name,'">',gene_name ,"</a>")) %>%
  datatable(., filter = 'top', options = list(pageLength = 10,lengthMenu = list(c(10, 25, 50, 100, -1), list('10', '25', '50','100', 'All')) ), escape = FALSE)
```

# Indels
```{r printsummaryindels, echo = FALSE}
# All defaults
maf <- readr::read_delim(snakemake@params[["abs_indels"]], delim = "\t", guess_max = 10^6)
nindels <- length(unique(maf$Start_Position))
```

Total number of mutations: `r nindels`

## High impact
```{r, echo = FALSE}
filter(maf, str_detect(Consequence, "frameshift|stop") | IMPACT == "HIGH") %>%
    mutate(tVAF = t_alt_count / t_depth, nVAF = n_alt_count / n_depth) %>%
    dplyr::select(Hugo_Symbol, Chromosome, Start_Position, Reference_Allele, Variant_Type, Tumor_Seq_Allele1, Tumor_Seq_Allele2, Consequence, IMPACT, tVAF, nVAF) %>%
    dplyr::rename(chrom = Chromosome, coord = Start_Position, gene_name = Hugo_Symbol) %>%
    mutate(gene_name = paste0('<a href="',"https://www.genecards.org/cgi-bin/carddisp.pl?gene=",gene_name,'">',gene_name ,"</a>")) %>%
    dplyr::select(chrom, coord, gene_name, dplyr::everything()) %>%
    datatable(filter = 'top', options = list(pageLength = 10, lengthMenu = list(c(10, 25, 50, 100, -1), list('10', '25', '50','100', 'All'))), escape = FALSE)
```

# CN plots

## CN plot
```{r CNplot, echo = FALSE}
# All defaults
include_graphics(snakemake@params[["abs_CNplot"]])
```

## BAF plot
```{r BAFplot, echo = FALSE}
# All defaults
```

# Structural variation

## Distribution of rearrangement types
```{r breakpointdistribution, echo = FALSE}
# All defaults
include_graphics(snakemake@params[["abs_breakpoint_dist"]])
```
